File: JavaScript/DP - 1D/198. House Robber.js
/* 198. House Robber
Difficulty: Medium
Submission Link: https://leetcode.com/problems/house-robber/submissions/1153674750/
*/

/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
    let currMax = 0
    let prevMax = 0

    for(let i = 0; i < nums.length; i++){
        let temp = currMax

        currMax = Math.max(prevMax + nums[i], currMax)

        prevMax = temp
    }
    return currMax
};

File: C#/DP - 1D/198. House Robber.cs
/* 198. House Robber
Difficulty: Medium
Submission Link: https://leetcode.com/problems/house-robber/submissions/1153674043/ */

public class Solution {
    private int[] houseMoney;
    private Dictionary<int, int> cache;

    public int Rob(int[] nums) {
        houseMoney = nums;
        cache = new Dictionary<int, int>();
        int n = houseMoney.Length;

        if (n == 1)
            return houseMoney[0];

        int maxAmountAtFirstPos = GetMaxAmount(0);
        int maxAmountAtSecondPos = GetMaxAmount(1);

        return Math.Max(maxAmountAtFirstPos, maxAmountAtSecondPos);
    }

    private int GetMaxAmount(int position) {
        if (cache.ContainsKey(position))
            return cache[position];

        int n = houseMoney.Length;

        if (position + 2 >= n)
            return houseMoney[position];

        if (position + 3 >= n) {
            cache[position] = houseMoney[position] + GetMaxAmount(position + 2);
            return cache[position];
        }

        int maxAmountSkippingOne = GetMaxAmount(position + 2);
        int maxAmountSkippingTwo = GetMaxAmount(position + 3);
        cache[position] = houseMoney[position] + Math.Max(maxAmountSkippingOne, maxAmountSkippingTwo);
        return cache[position];
    }
}